Update Joomla Application packages to framework v2

wilsonge opened the pull request:
### Summary of Changes
Updates the Joomla Application package to version 2 of the framework. 

Also phpmailer 6.0rc4 from rc3 as well as updating the dependencies of some of our test apps

### Testing Instructions
Ensure the Joomla v4 branch has no additional breaks

### Documentation Changes Required
Document the new version of packages included

wilsonge committed:
Update Joomla Application, Registry and Data packages to framework v2

mbabker commented:
IMO we should go through the Framework repos first and start marking cross-version compatibility between the packages (most of them are either strictly 1.x or 2.x, not 1.x & 2.x where able).  That way we can better test the inner dependencies then start working our way out.

But yes, this should be done soon.

mbabker commented on libraries/joomla/application/web.php:
```diff
@@ -116,9 +116,9 @@
-		$this->client = $client ?: new WebClient;
+		$input = $input ?: new JInput;
+
+		parent::__construct($input, $config, $client);
```
We should allow for the PSR-7 object injection here too.  Even if the CMS itself won't actually use it, if someone creates a custom subclass from here they may be more interested in that feature.  I don't think we'll need it on `JApplicationCms` or its subclasses though.

mbabker commented:
You've got the session 1.0 package installed here too.

wilsonge commented:
Removed session tests, added the DI as requested :)

wilsonge committed:
Add method to tests

joomdonation commented:
I have tested this item :white_check_mark: successfully on 4fbba2e85db81d2bf14408f640b3dea246589f08<br /><br />Installed a copy of https://github.com/wilsonge/joomla-cms/tree/update-application with sample data. Created user account, articles, works

Tried to login from frontend, works. Tried to navigate through available menu items, found fatal error with some of them but it even happens before this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14145">issues.joomla.org/tracker/joomla-cms/14145</a>.</sub>

wilsonge committed:
Try and fix tests

wilsonge commented:
I have removed tests for `testAppendBody`, `testPrependBody`, `testSetBody` and `testClearHeaders`, `testGetHeaders`, `testSendHeaders` and `testSetHeader` as these are no longer set in `JApplicationWeb` but instead at the framework level and therefore should be tested in the framework. I've fixed the tests for `render` in all 4 locations

wilsonge committed:
Add check token proxy to CMS JSession::checkToken

wilsonge commented:
I've just added a method that overrides the checkToken method in `AbstractWebApplication` and proxies to `JSession::checkToken` so we have one method that covers all our checks

C-Lodder commented:
There are lots of instances in this PR that define an array using `[]` as opposed to `array()`. This is a PHP 5.4 feature and J3 still supports PHP 5.3.

wilsonge commented:
This is a Joomla 4 PR ;)

C-Lodder commented:
Ahh, didn't see the branch, just looked at the title.
